R/add_subprotocol.R

```diff
@@ -0,0 +1,220 @@
+      path_rel(start = find_root(is_git_root))
+
+    tag <- paste(code_subprotocol, version_number, sep = "-")
+    gitcommand <- paste0("git show ",
+                         tag, ":",
+                         git_filepath)
+
+    # #Fetching the git repo
+    # # this usually returns "origin"
+    # firstremote <- execshell("git remote", intern = TRUE)[1]
+    # execshell(paste0("git fetch ", firstremote))
+
+    # get the content of the Rmd file
+    # this will return a character vector (each element is one sentence)
+    rmd_content <- execshell(gitcommand,
+           intern = TRUE)
+    # What happens if this fails?
+    # for instance because protocol_code and version_number don't match
+
+    # handling the section arguments
+    # avoid looking in chunks which can have lines starting with '#'
+    is_chunk <- grepl("^`{3}", rmd_content)
+    i <- 1
+    while (i < length(is_chunk)) {
+      if (isTRUE(is_chunk[i]) && isTRUE(is_chunk[i+1])) {
+        i <- i + 2
+      } else if (isTRUE(is_chunk[i]) && isFALSE(is_chunk[i+1])) {
+        is_chunk[i+1] <- TRUE
+        i <- i + 1
+      } else {
+        i <- i + 1
+      }
+    }
+
+    if (!missing(section)) {
+      has_section <- grepl(section, rmd_content, fixed = TRUE)
+
+      # assuming section is header level 2
+      h2 <- grepl("^##\\s[A-Z]", rmd_content) & !is_chunk
+      # grab the section
+      start_section <- which(has_section)
+      if (length(start_section) > 1) {
+        stop(
+          "section does not uniquely identify one section"
+        )
+      }
+      h2_rows <- which(h2)
+      end_section <- min(h2_rows[h2_rows > start_section]) - 1
+      rmd_content <- rmd_content[start_section:end_section]
+      is_chunk <- is_chunk[start_section:end_section]
+    }
+    # demote headers
+    if (demote_header != 0) {
+      h1 <- grepl("^#\\s[A-Z]", rmd_content) & !is_chunk
+      h2 <- grepl("^##\\s[A-Z]", rmd_content) & !is_chunk
+      h3 <- grepl("^###\\s[A-Z]", rmd_content) & !is_chunk
+      if (demote_header == -1) {
+        if(any(h1)) {
+          stop("demote header -1 not possible when a level 1 header is present")
+        } else {
+          rmd_content[h2] <- str_remove(rmd_content[h2], "^#")
+          rmd_content[h3] <- str_remove(rmd_content[h3], "^#")
+        }
+      } else {
+        rmd_content[h1] <- paste0(
+          paste0(rep("#", demote_header), collapse = ""), rmd_content[h1])
+        rmd_content[h2] <- paste0(
+          paste0(rep("#", demote_header), collapse = ""), rmd_content[h2])
+        rmd_content[h3] <- paste0(
+          paste0(rep("#", demote_header), collapse = ""), rmd_content[h3])
+      }
+    }
+
+
+    # insert a yaml section with params name-value pairs
+    if (!missing(params)) {
+      pairs <- sprintf("  %s: %s", names(params), params)
+      rmd_content <- c("---",
+                       "params:",
+                       pairs,
+                       "---",
+                       rmd_content)
+    }
```

Comment:
Seems a bit odd to me. The function inserts one file - hence one chapter - of another protocol into the overall Rmd of a project protocol. But the subprotocol has its `params` defined (with defaults) in its own `index.Rmd`. I rather not expect YAML sections within an Rmd to work (i.e. at the beginning of an appendix, in this case). Furthermore, setting `params` when knitting is done through `rmarkdown::render()`.

So probably two things are needed to correctly set `params$xxx` with this function:
- prepending a `setup-inserted-section` R chunk to the imported Rmd code (instead of a YAML header), in which the `params` object is created (empty list) and elements defined.
- searching for the defaults of the other `params` elements in the `index.Rmd` of the subprotocol, and adding these in the `setup-inserted-section` chunk.

Reply:
Which explains your earlier comment:

> the inserted yaml params section does not seem to overwrite existing params defined in index.Rmd

Reply:
Now I now where I read that multiple yaml sections are possible: https://pandoc.org/MANUAL.html#metadata-blocks:
> A document may contain multiple metadata blocks. If two metadata blocks attempt to set the same field, the value from the second block will be taken.

That was why I thought the above approach would work. But it didn't. I have also tried the approach of defining the params in an R code chunk as you suggested, but various attempts also failed (I could not overwrite the locked bindings). So I've abandoned that idea too and will now only focus on approaches that pass the non-default params via `render()` (which is the recommended approach).

Reply:
OK Hans, interesting, thanks for looking into it and for this update!

Reply:
I think I might see what could have gone wrong. The inserted yaml section does (see a few posts above) does not guarantee that a white line is before and after the `---`. This may have caused pandoc to not recognize it as yaml block...
I might reconsider this approach...

Reply:
I read that too. But actually I would be quite surprised that that works. `pandoc` is called to process the `md` file, i.e. only after `knit()` has run the code chunks and generated markdown output for each of them (the Rmd file is converted to a temporary or preserved md file). Hence `pandoc` just processes the entire resulting markdown file; it would read those `params` keys and extra blocks but it would not do anything with them as they are for use by `knit()` in handling R chunks.

Reply:
Indeed. That surely explains why it didn't work.